In [14]:
import numpy as np
import pandas as pd

df = pd.read_csv('merged_adf_df_with_lda_topic1.csv', index_col=0)

In [5]:
'''from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Embedding, LSTM, SpatialDropout1D
from sklearn.model_selection import train_test_split
from keras.utils.np_utils import to_categorical

# Tokenize the text
tokenizer = Tokenizer(num_words=5000, split=' ')
tokenizer.fit_on_texts(df['LemmatizedQuestionBody'].values)

# Convert text to sequences
X = tokenizer.texts_to_sequences(df['LemmatizedQuestionBody'].values)
X = pad_sequences(X)

# Build the model
embed_dim = 128
lstm_out = 196

model = Sequential()
model.add(Embedding(5000, embed_dim, input_length = X.shape[1]))
model.add(SpatialDropout1D(0.4))
model.add(LSTM(lstm_out, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(2, activation='softmax')) # 2 for binary classification
model.compile(loss = 'categorical_crossentropy', optimizer='adam', metrics = ['accuracy'])

# Split the data into training and testing sets
Y = pd.get_dummies(df['LDATopic']).values
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 42)

# Train the model
batch_size = 32
model.fit(X_train, Y_train, epochs = 7, batch_size=batch_size)

# Evaluate the model
score, acc = model.evaluate(X_test, Y_test, verbose = 2, batch_size = batch_size)
print("score: %.2f" % (score))
print("acc: %.2f" % (acc))'''

Epoch 1/7
214/214 [==============================] - 4861s 23s/step - loss: 0.4608 - accuracy: 0.7889
Epoch 2/7
214/214 [==============================] - 4851s 23s/step - loss: 0.3037 - accuracy: 0.8660
Epoch 3/7
214/214 [==============================] - 4776s 22s/step - loss: 0.2250 - accuracy: 0.9084
Epoch 4/7
214/214 [==============================] - 4955s 23s/step - loss: 0.1712 - accuracy: 0.9349
Epoch 5/7
214/214 [==============================] - 4744s 22s/step - loss: 0.1243 - accuracy: 0.9522
Epoch 6/7
214/214 [==============================] - 5420s 25s/step - loss: 0.0920 - accuracy: 0.9650
Epoch 7/7
214/214 [==============================] - 5518s 26s/step - loss: 0.0714 - accuracy: 0.9729
54/54 - 72s - loss: 0.5112 - accuracy: 0.8684 - 72s/epoch - 1s/step
score: 0.51
acc: 0.87


In [9]:
# Save the model
#model.save('RNN_model.h5')  # HDF5 file


***GPT4 prompt***: How to interpret these results?

***GPT4 result***:

The results you posted are from training and evaluating a Long Short-Term Memory (LSTM) model with TensorFlow. LSTM is a type of Recurrent Neural Network (RNN) that's good for sequence data, like text.

The training part of your model is done over 7 epochs. Each epoch is a full pass through your dataset. Here's a summary of what happens in each epoch:

    Epoch 1: The model starts with a loss of 0.4608 and accuracy of 78.89%. Loss is a measure of error—lower is better. Accuracy is the percentage of correct predictions—higher is better. The model is learning, but it's just starting, so it's not very accurate yet.

    Epoch 2-6: The loss keeps decreasing and the accuracy keeps increasing. This is a good sign. It means the model is learning from the data and getting better at making predictions.

    Epoch 7: The model ends with a loss of 0.0714 and accuracy of 97.29%. This is a high accuracy and low loss, suggesting that the model is performing well on the training data.

After training, you evaluate the model on your test data. This is data that the model has never seen before, which gives you a good idea of how the model will perform on new, unseen data. The model ends with a test score (loss) of 0.51 and test accuracy of 87%, which is quite good.

However, there's a significant difference between the accuracy on the training data (97.29%) and the accuracy on the test data (87%). This suggests that the model may be overfitting to the training data, i.e., it's learning the training data so well that it performs poorly on new data. You may need to adjust your model or add regularization to reduce overfitting.

In summary, you've built a LSTM model that's achieving high accuracy on your training data, but it might be overfitting. You might need to try different strategies (like adding dropout, trying a simpler model, or collecting more data) to reduce overfitting and improve performance on your test data.

In [15]:
from tensorflow.keras.models import load_model

model = load_model('RNN_model.h5')

In [16]:
from sklearn.metrics import confusion_matrix, classification_report
import matplotlib.pyplot as plt
from keras.utils import plot_model

# Predict the values from the validation dataset
Y_pred = model.predict(X_test)
# Convert predictions classes to one hot vectors 
Y_pred_classes = np.argmax(Y_pred, axis = 1) 
# Convert validation observations to one hot vectors
Y_true = np.argmax(Y_test, axis = 1)
# compute the confusion matrix
confusion_mtx = confusion_matrix(Y_true, Y_pred_classes)

print('Confusion Matrix')
print(confusion_mtx)

54/54 [==============================] - 46s 844ms/step
Confusion Matrix
[[ 294  118]
 [ 107 1191]]
